File: 1819-2/nonprocedural_programming/random/bfs.pl
% bfs(Edges, Queue, Seen)
bfs(_, [], _, []) :- !.
bfs(Edges, [Q|Queue], Seen, [Q|Path]) :-
    findall(A, 
            (
                member(Q-A, Edges),
                \+ member(A, Seen),
                \+ member(A, Queue)
            ),
            ToAdd
            ),
    append(Queue, ToAdd, NextQueue),
    bfs(Edges, NextQueue, [Q|Seen], Path).

e1([a-b, a-c, b-c, c-d, b-e]).

File: 1819-2/nonprocedural_programming/random/bfs.hs
type Edges = [(Char, Char)]
type Verticies a = [(Char, a)]
data Graf a= EV (Edges, Verticies a) deriving (Show)

-- Edges, Queue, Seen
bfs :: Graf a -> [Verticies a] -> [Verticies a] -> Int -> Graf (a, Int)
bfs _ [] seen _ = []
bfs (EV (edges, verticies)) queue seen num = (EV (edges, new_verticies++bfs_verticies)) 
    where
        (EV (_, new_verticies)) = bfs (EV (edges, verticies)) next_queue (seen ++ queue) (num+1)
        next_queue = foldl (++) [] [neighbours edges x| x <- queue]
        